### PR TITLE
[Java API] Provide byte[] version of SstFileWriter.merge to reduce GC Stall

### PR DESCRIPTION
**Pull request by scv119**
In Java API, `SstFileWriter.put/merge/delete` takes `Slice` type of key and value, which is a Java wrapper object around C++ Slice object.  The Slice object inherited [ `finalize`](https://github.com/facebook/rocksdb/blob/3c327ac2d0fd50bbd82fe1f1af5de909dad769e6/java/src/main/java/org/rocksdb/AbstractNativeReference.java#L69) method, which [added huge overhead](https://softwareengineering.stackexchange.com/questions/288715/is-overriding-object-finalize-really-bad/288753#288753) to JVM while creating new SstFile.

To address this issue, this PR overload the merge method to take Java byte array instead of the Slice object, and added unit test for it.

We also benchmark these two different merge function, where we could see GC Stall reduced from 50%  to 1%, and the throughput increased from 50MB to 200MB.

### CODE REVIEW

File: java/rocksjni/sst_file_writerjni.cc [hunk 1]
```diff
@@ -110,6 +110,43 @@ void Java_org_rocksdb_SstFileWriter_merge(JNIEnv *env, jobject jobj,
...
+                                                   jlong jhandle, jbyteArray jkey,
+                                                   jbyteArray jval) {
+
+  jbyte* key = env->GetByteArrayElements(jkey, nullptr);
+  if(key == nullptr) {
+    // exception thrown: OutOfMemoryError
+    return;
+  }
+  rocksdb::Slice keySlice(
+      reinterpret_cast<char*>(key),  env->GetArrayLength(jkey));
```
Comment (sagar0) [inline java/rocksjni/sst_file_writerjni.cc]: It seems like the array lengths are also passed in for some of the other native method implementations. For examples, see: https://github.com/facebook/rocksdb/blob/master/java/src/main/java/org/rocksdb/RocksDB.java#L2257-L2268 . 

I am not aware of the difference between passing in the array length explicitly vs. using env-> GetArrayLength() method used here and in line 134. Also, those functions (in the example I pointed out above) seem to be taking an 'offset' as well, but it is always set to 0. Not sure use of the usefulness of offset either. 

If a length and offset need to be passed in, the function signature needs to change to ...merge_J_3BII_3BII . 

@adamretter could you elaborate on the usefulness of passing in length and offset vs. using env->GetArrayLength()?

**Comment by adamretter** [comment]
@scv119 I can see why you proposed the changes you have when using Slice. However, why not just use `DirectSlice`?  You can update the `ByteBuffer` behind a `DirectSlice`, and the C++ `Slice` will see the updated data without you needing to do anything else. So you could have a single `DirectSlice`, which you use over and over again.

File: java/rocksjni/sst_file_writerjni.cc (hunk 1 above)
Comment (adamretter) [inline java/rocksjni/sst_file_writerjni.cc]: @sagar0 Two reasons:

1. Passing in the `length` is likely cheaper than calling `env->GetArrayLength`
2. Passing in the `length` and `offset` allow you to just store part of an existing byte array. Users have previously requested being able to store part of an array without them needing to copy to a smaller array first. This makes sense when they have many keys/values encoded into a single byte array.

File: java/rocksjni/sst_file_writerjni.cc [hunk 2]
```diff
@@ -92,12 +92,49 @@ void Java_org_rocksdb_SstFileWriter_put(JNIEnv *env, jobject jobj,
...
+    // exception thrown: OutOfMemoryError
+    return;
+  }
+  rocksdb::Slice key_slice(
+      reinterpret_cast<char*>(key),  env->GetArrayLength(jkey));
+
+  jbyte* value = env->GetByteArrayElements(jval, nullptr);
+  if(value == nullptr) {
+    // exception thrown: OutOfMemoryError
+    env->ReleaseByteArrayElements(jkey, key, JNI_ABORT);
```
Comment (scv119) [inline java/rocksjni/sst_file_writerjni.cc]: I'm not verify familiar how exception is thrown here but I guess I still need to release the key related pointers.

**Comment by scv119** [comment]
@adamretter Ah I just noticed we have a DirectSlice, I could give it a try and see how does it perform. 
At the mean time do you think this PR still worth as it unify the JNI interface? So that `RocksDB.write/merge`, `RocksIterator.seek` and `SstFileWriter.put/merge/delete` could all use byte[] type API?

**Comment by scv119** [comment]
After reading the DirectSlice code and think I might stick to byte[] version for following reasons
1. As I mentioned previously, it's worth to unify the java api to use byte[], as the rocksdb.merge/put/get already uses byte[] instead of slice.
2. Reuse DirectSlice could solve the problem, but I need to reuse the direct buffer, (if I understand correctly), so I have to allocate enough space and also write some code to replace the content in the bytebuffer. This seems a bit hacky to me and makes code less readable.
What do you think?

**Pull request closed without merging**